src/ucp/proto/proto_perf.c

```diff
@@ -231,16 +224,16 @@ ucs_status_t ucp_proto_perf_from_caps(const char *name,
         ucp_proto_perf_node_ref(seg->node);
         ucs_list_add_tail(&perf->segments, &seg->list);
 
-        ucp_proto_perf_segment_add_func(perf, seg, UCP_PROTO_PERF_FACTOR_SINGLE,
-                                        range->perf[UCP_PROTO_PERF_TYPE_SINGLE],
-                                        NULL);
-        ucp_proto_perf_segment_add_func(perf, seg, UCP_PROTO_PERF_FACTOR_MULTI,
-                                        range->perf[UCP_PROTO_PERF_TYPE_MULTI],
-                                        NULL);
-        ucp_proto_perf_segment_add_func(perf, seg,
-                                        UCP_PROTO_PERF_FACTOR_LOCAL_CPU,
-                                        range->perf[UCP_PROTO_PERF_TYPE_CPU],
-                                        NULL);
+        /* funcs array should be according to the order of factor_id bits */
```

Comment:
to remove that constraint, what about instead simply pass `funcs` being defined as
```
ucs_linear_func_t funcs[UCP_PROTO_PERF_FACTOR_LAST] = {
  [UCP_PROTO_PERF_TYPE_CPU] = range->perf[UCP_PROTO_PERF_TYPE_CPU],
...
};
```

and either only use non-null pointers or keep the bitfield?

Reply:
then if the user wants to define few funcs, they'd still have to allocate an array of the maximal length.

Reply:
But each segment already contains all the funcs and even if it would be used explicitly (not during aggregation of two perf structures) that funcs will be probably allocated on stack (and even optimized-away by compiler) so is that a significant drawback?

Reply:
IMO the drawback is to initialize an array with NULL pointers / "empty" values. I guess we can use a bitmap to say which fields are valid.

Reply:
fixed